### ISSUE TITLE
Some values are encoded in a non-readable way

### ISSUE THREAD

**Issue by captain-majid** [icza/screp#59]
```
{
        "Frame": 444,
        "PlayerID": 0,
        "Type": {
          "Name": "Cheat",
          "ID": 18
        },
        "Data": "AQAAAA=="
},
```

Would like to know the cheat type, so how **Data** is encoded here ?

**Comment by icza** [comment]
JSON encodes binary data as Base64.

**Comment by captain-majid** [comment]
I tried to convert that to byte array which gives `[1 0 0 0]` but the value isn't static, I think it depends on whether other cheats ran with it or b4 it or not, probabply using enum bit-or to represent the whole state of cheat codes, any idea how this is strctured ? otherwise the "Data": "AQAAAA==" isn't useful.

**Comment by icza** [comment]
Binary data of some commands are not parsed / modeled. They are still retained and outputted as-is, which reflects the data stored in replays. In such cases the parser needs to be updated to interpret the data, or you have to do your own research what the data means.

**Comment by captain-majid** [comment]
Any othe values are still encoded as base64 string in json ?

**Comment by icza** [comment]
Yes, there are some, there are also some that partly interpret the data, and outputs the rest, uninterpreted part as Base64.

**Comment by captain-majid** [comment]
cmds like what exactly ?

**Comment by icza** [comment]
Actually, looking at the code, I don't see anything else besides the "Use Cheat" command.

**Comment by captain-majid** [comment]
I think I finally got it working, this code is in C#, no experience in Go, so hopefully you can port it and tell me if it's working on your side:
```cs
if (cmdType == "Cheat")
{
    cheatByteArray_new = Convert.FromBase64String(item.XPathSelectElement("Data").Value); // convert encoded Base64 string to byte array
    string[] excludeThose = new string[] { "16,0,0,0" , "0,0,4,0", "0,0,8,0", "8,0,0,0" }; //don't know how, but ignoring those seems to have the correct values.
    byte[] result = new byte[4];
    for (int i = 0; i < 4; i++)
    {
        result[i] = (byte)(cheatByteArray_new[i] ^ cheatByteArray_old[i]); //xoring the items in the 2 byte arrays
    }
    data = string.Join(",", result);
    if (!excludeThose.Contains(string.Join(",", result)))
        cheatByteArray_old = cheatByteArray_new;

    switch (data) //The initial values of every cheat code, can be different if the order of issuing them changed.
    {
        case "16,0,0,0": more = ": show me the money " + data; break;
        case "1,0,0,0": more = ": black sheep wall " + data; break;
        case "2,0,0,0": more = ": operation cwal " + data; break;
        case "4,0,0,0": more = ": power overwhelming " + data; break;
        case "0,1,0,0": more = ": staying alive " + data; break;
        case "128,0,0,0": more = ": there is no cow level " + data; break;

        case "64,0,0,0": more = ": game over man " + data; break;
        case "0,0,4,0": more = ": whats mine is mine " + data; break;
        case "0,8,0,0": more = ": the gathering " + data; break;
        case "0,0,8,0": more = ": breathe deep " + data; break;

        case "8,0,0,0": more = ": something for nothing " + data; break;
        case "0,16,0,0": more = ": medieval man " + data; break;
        case "0,32,0,0": more = ": modify the phase variance " + data; break;
        case "0,64,0,0": more = ": war aint what it used to be " + data; break;

        case "0,0,2,0": more = ": food for thought " + data; break;
        case "0,0,0,32": more = ": noglues " + data; break;
        case "0,2,0,0": more = ": ophelia " + data; break;
        case "0,0,16,0": more = ": zoom zoom " + data; break;
        default: more = ": ??? unknown " + data; break;
    }
}
```

**Comment by icza** [comment]
So it seems there are 4 bytes of data, and each bit of those corresponds to a specific cheat code (there may be unsued bits).

**Comment by captain-majid** [comment]
I found a page that clarify more how these cheat codes are encoded, it also contain many other orders that might not be identified by Screp:
https://github.com/phu54321/vgce/blob/0da9d14626c06351e66a3d9c18973f6d69b89a7a/docs/Blizzard/Starcraft/packets2.txt#L662

**Comment by icza** [comment]
Thanks for the link, I will implement parsing cheat codes.

**Comment by icza** [comment]
And there ya go: released v1.13.0 that parses which cheats are enabled / were issued: https://github.com/icza/screp/releases/tag/v1.13.0

**Comment by captain-majid** [comment]
Thanks, but the way those cheats are encoded is intriguing, they don't really have a fixed id unless you use 1 cheat in a game, if a cheat used multiple times or along with other cheats, you won't get the same signature/id of the cheat, this is why I used `cheatByteArray_old` and `cheatByteArray_new` in my C# code to be aware of what was the previous cheat and `xor` it with the new1.

I use the following sample replay to test if cheats are recognized correctly or not, and unfortunately screp v1.13 doesn't match what really happened, I also don't understand the necessity of "CheatCodes": [] in the generated .json file, its only 1 cheat at time, so why an array that still repeat all the previous cheats in it again?

[! Omega_test cheat codes.zip](https://github.com/user-attachments/files/23967141/Omega_test.cheat.codes.zip)

**Comment by icza** [comment]
The cheat command sends a bitmap that also holds cheats that are toggleable and are currently turned on.

E.g. "Black sheep wall" (reveals full map) is a toggleable cheat, onces enabled, you'll see the full map. If you enter it again, it disables it.

There are other, non-toggleable cheats, e.g. "Show me the money" which gives you 10k minerals and gas on each use.

The `[]*CheatCodes` slice holds cheats decoded from the bitmap: they are either turned on toggleable cheats, or a repeatable cheat that was just used.

I see no issues here.

**Comment by captain-majid** [comment]
Look for example @frame 4520 in the replay I've encluded, by looking at the .json file I don't really know what cheat was that:
```json
      {
        "Frame": 4520,
        "PlayerID": 0,
        "Type": {
          "Name": "Cheat",
          "ID": 18
        },
        "CheatsBitmap": 537555456,
        "CheatCodes": [
          {
            "Name": "Ophelia",
            "BitMask": 512
          },
          {
            "Name": "Medieval Man",
            "BitMask": 4096
          },
          {
            "Name": "Modify The Phase Variance",
            "BitMask": 8192
          },
          {
            "Name": "War Aint What it Used to be",
            "BitMask": 16384
          },
          {
            "Name": "Food For Thought",
            "BitMask": 131072
          },
          {
            "Name": "Breathe Deep",
            "BitMask": 524288
          },
          {
            "Name": "Noglues",
            "BitMask": 536870912
          }
        ]
      },
```

But I know that exactly in this frame I just entered `Cheat: breathe deep` with no other cheats !

**Comment by icza** [comment]
Again: the Cheat command reports the state of all toggleable cheat codes.

Amongst all the listed cheat codes, only "breathe deep" is non-toggleable, which you just entered. All the other cheat codes are toggleable cheats which according to this command are all turned on. Note that some cheats persist state between games, such as "Ophelia" or "Noglues".

**Comment by captain-majid** [comment]
You mean with every cheat entered in the game, I get a list of all the enabled cheats (and also the repeatable ones if any), this is complex enough to make you don't want to know about the cheats entered at all.

I can't think of any easy way to just get cheat that was entered in the frame from the current screp version, the previous version which encoded them as Base64string worked great with me. I don't know what to do now.

**Comment by icza** [comment]
The previous Base64String data is still available and there: it's the `CheatsBitmap` field in the JSON, just not as the base64 encoded value of the 4 bytes data, but rather as the interpreted unsigned 32-bit integer value.

The `CheatCodes` doesn't list all issued cheat commands. If you enter "Black sheep wall", you'll see only that. If you enter it again, you'll see "Cheats Disabled". If you enter "show me the money", you'll see only that. If you enter it again, you again will only see that.

You only see multiple cheats, if there are some that are toggleable and are turned on. This is how it is recorded in replays. This is basically the "cheat state".
